File: src/main/java/com/team3/gudit/auth/oauth2/KakaoUserInfo.java
package com.team3.gudit.auth.oauth2;

import java.util.Map;

public record KakaoUserInfo(
        Map<String, Object> attributes
) implements OAuth2UserInfo {

    @Override
    public Long id() {
        Object id = attributes.get("id");
        return id == null ? null : Long.valueOf(id.toString());
    }

    @Override
    public String email() {
        Map<String, Object> kakaAccount = kakaAccount();
        return kakaAccount == null ? null : String.valueOf( kakaAccount.get("email") );
    }

    @Override
    public String name() {
        Map<String, Object> profile = profile();
        return profile == null ? null : String.valueOf( profile.get("nickname") );
    }

    @Override
    public String imageUrl() {
        Map<String, Object> profile = profile();
        return profile == null ? null : String.valueOf( profile.get("profile_image_url") );
    }

    private Map<String, Object> kakaAccount() {
        return (Map<String, Object>) attributes.get("kakao_account");
    }

    private Map<String, Object> profile() {
        return (Map<String, Object>) kakaAccount().get("profile");
    }

}

File: src/main/kotlin/com/team3/gudit/auth/oauth2/KakaoUserInfo.kt
package com.team3.gudit.auth.oauth2

@JvmRecord
data class KakaoUserInfo(private val attributes: Map<String, Any>) : OAuth2UserInfo {
    override fun attributes(): Map<String, Any> = attributes
    override fun id(): Long? = attributes["id"]?.toString()?.toLong()

    override fun email(): String? {
        val account = kakaoAccount()
        return account?.get("email")?.toString()
    }

    override fun name(): String? {
        val profile = profile()
        return if (profile == null) null else profile["nickname"].toString()
    }

    override fun imageUrl(): String? {
        val profile = profile()
        return if (profile == null) null else profile["profile_image_url"].toString()
    }

    @Suppress("UNCHECKED_CAST")
    private fun kakaoAccount(): Map<String, Any>? = attributes["kakao_account"] as Map<String, Any>?

    @Suppress("UNCHECKED_CAST")
    private fun profile(): Map<String, Any>? = requireNotNull(kakaoAccount())["profile"] as Map<String, Any>?
}
